### PR TITLE
Emails campaign detection script

### PR DESCRIPTION
**Pull request by lperryp**
## Status
- [ ] In Progress
- [ ] Ready
- [x] In Hold - (waiting for this PR to complete: https://github.com/demisto/content/pull/10762)

## Related Issues
fixes: https://github.com/demisto/etc/issues/32643

## Description
A new script for characterizing phishing emails campaigns. 
The script finds phishing incidents with high text similarity, and returns some summarization of them, such as: 

1. Details (date range, involved senders, similarity degree)
2. Summarization of the email using heuristic + highlighting of spam/malicious keywords
3. A table of mutual indicators + their reputation
4. A table of relevant incidents

## Screenshots
Paste here any images that will help the reviewer
image
image

image

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by lperryp** [comment]
@erezh31 - opened a PR for the campaign script so it'd be possible to review it, notice that only FindEmailsCampaign.py + FindEmailsCampaign.yml are relevant here, the rest were already reviewed here:  https://github.com/demisto/content/pull/10762

File: Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    EMAIL_HTML_FIELD = input_args.get('emailBodyHTML', EMAIL_HTML_FIELD)
+    FROM_FIELD = input_args.get('emailFrom', FROM_FIELD)
+
+    res = demisto.executeCommand('FindDuplicateEmailIncidents', input_args)
+    if is_error(res):
+        return_error(get_error(res))
+    res = res[-1]
+    incidents = json.loads(res['Contents'])
+    if is_number_of_incidents_is_to_low(res, incidents):
+        return
```
Comment (erezh31) [inline Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.py:328]: *too

File: Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        axis=1, inplace=True)
+    incidents_headers = ['Id', 'Created', 'Name', 'Email From', 'Similarity to Current Incident']
+    hr = '\n\n' + tableToMarkdown('Involved incidents', incidents_df[incidents_headers].to_dict(orient='records'),
+                                  headers=incidents_headers)
+    return_outputs(hr)
+
+
+def analyze_incidents_campaign(incidents_df):
+    return_campaign_details_entry(incidents_df)
+    return_indicator_entry(incidents_df)
```
Comment (erezh31) [inline Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.py:310]: return here the indicators df from this fucntion

File: Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    return_involved_incdients_entry(incidents_df)
```
Comment (erezh31) [inline Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.py:311]: return the incidents df from this function

File: Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+    cv = CountVectorizer(stop_words=list(stopwords.words('english')))
+    body_arr = cv.fit_transform(corpus).toarray()
+    subject_arr = cv.transform(sent_tokenize(subject)).toarray()
+    word_list = cv.get_feature_names()
+    count_list = body_arr.sum(axis=0) + subject_arr.sum(axis=0) * 1.5
+    duplicate_sentences = [i for i, arr in enumerate(body_arr) if
+                           any(cosine_sim(arr, arr2) > DUPLICATE_SENTENCE_THRESHOLD
+                               for arr2 in body_arr[:i])]
+
+    """
```
Comment (erezh31) [inline Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.py:191]: remove this

File: Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        domain_header = "Sender domain"
+        sender_header = "Sender address"
+
+    elif len(senders_counter) > 1 and len(domains_counter) == 1:
+        domain_header = "Senders domain"
+        sender_header = "Senders addresses"
+    else:
+        domain_header = "Senders domains"
+        sender_header = "Senders addresses"
+    top_n = 1
```
Comment (erezh31) [inline Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.py:164]: maybe more? 3?

File: Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.yml [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  name: statusScope
+  predefined:
+  - All
+  - ClosedOnly
+  - NonClosedOnly
+  required: false
+  secret: false
+- default: false
+  defaultValue: '0.8'
+  description: Threshold to consider incident as duplication, number between 0-1
```
Comment (erezh31) [inline Packs/Phishing/Scripts/FindEmailsCampaign/FindEmailsCampaign.yml:90]: I would change the description Use similar and not duplicate

**Review by erezh31 — CHANGES_REQUESTED** [review]

**Comment by idovandijk** [comment]
Looks really useful, thanks. I have additional points from observing on our server:
1. Can we have the involved incident count show in context output as well?
2. Something is a bit off with the context. Why is there [0] under EmailCampaign, and then 0, 1, 2, 3, 4? Under 1 I see incidents, but under the rest I see only indicators. What does each one represent? Under incidents there is also 0, 1, 2, 3, 4 which is the one that does make sense to me
image
3. Can we also output to context the sender domains and addresses and the list of recipients? I'm not sure we need it under indicators, but rather under their corresponding sections. This way we can differentiate more easily between indicators related to us and indicators related to the attackers

File: Packs/Phishing/Scripts/FindEmailCampaign/FindEmailCampaign.py [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+    for i, sent in enumerate(corpus):
+        if i in duplicate_sentences:
+            continue
+        sentence_rank[i] = 0
+        for word in word_tokenize(sent):
+            if word.lower() in word_frequency.keys():
+                if i in sentence_rank.keys():
+                    sentence_rank[i] += word_frequency[word.lower()]
+        sentence_rank[i] = sentence_rank[i] / len(word_tokenize(sent))  # type: ignore
+    sorted_sentence_rank = sorted(sentence_rank.items(), key=lambda item: item[1], reverse=True)
```
Comment (erezh31) [inline Packs/Phishing/Scripts/FindEmailCampaign/FindEmailCampaign.py:264]: I think there is numpy function that will return you the indices directly, like argmax or something

File: Packs/Phishing/Scripts/FindEmailCampaign/FindEmailCampaign.py [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+        indicators_df.rename({'Value': 'value'}, axis=1, inplace=True)
+        indicators_df = indicators_df[['id', 'value']]
+        indicators_context = indicators_df.to_dict(orient='records')
+    return {'indicators': indicators_context}
+
+
+CONTEXT_FUNCTIONS = [create_context_for_campaign_details, create_context_for_indicators, ]
+
+
+def create_empty_context():
```
Comment (erezh31) [inline Packs/Phishing/Scripts/FindEmailCampaign/FindEmailCampaign.py:113]: usually in this case we can just not reutrn a context, or to return the campaign idenfity false context key. why we need all the keys?

File: Packs/Phishing/Scripts/FindEmailCampaign/FindEmailCampaign.py [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+def is_number_of_incidents_too_low(res, incidents):
+    if not res["EntryContext"]['isDuplicateIncidentFound'] or \
+            len(incidents) < MIN_CAMPAIGN_SIZE:
+        return_outputs_custom('No possible campaign was detected', create_empty_context())
+        return True
+    return False
+
+
+def is_number_of_unique_recipients_is_too_low(incidents):
+    unique_recipients = Counter([str(i.get('emailto', 'None')) for i in incidents])
```
Comment (erezh31) [inline Packs/Phishing/Scripts/FindEmailCampaign/FindEmailCampaign.py:130]: I think you should check also CC, BCC fields to get all the recipients

File: Packs/Phishing/Scripts/FindEmailCampaign/FindEmailCampaign.py [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+        if i in duplicate_sentences:
+            continue
+        sentence_rank[i] = 0
+        for word in word_tokenize(sent):
+            if word.lower() in word_frequency.keys():
+                if i in sentence_rank.keys():
+                    sentence_rank[i] += word_frequency[word.lower()]
+        sentence_rank[i] = sentence_rank[i] / len(word_tokenize(sent))  # type: ignore
+    sorted_sentence_rank = sorted(sentence_rank.items(), key=lambda item: item[1], reverse=True)
+    top_sentences_indices = [sent_i for sent_i, _ in sorted_sentence_rank[:nb_sentences]]
```
Comment (erezh31) [inline Packs/Phishing/Scripts/FindEmailCampaign/FindEmailCampaign.py:265]: Make sure the display order of the senteces is the same as they appeared in the original email. Here I'm not sure if the order is by the rank you gave them

**Review by erezh31 — CHANGES_REQUESTED** [review]

File: Packs/Phishing/Scripts/FindEmailCampaign/FindEmailCampaign.py (hunk 10 above)
Comment (lperryp) [reply Packs/Phishing/Scripts/FindEmailCampaign/FindEmailCampaign.py:265]: they are sorted by order later

**Review by erezh31 — APPROVED** [review]

**Comment by meirwah** [comment]
@Michal-Vardi any ETA on the review here ?

**Comment by idovandijk** [comment]
@erezh31 / @lperryp could you please update the script on our server after the needed changes mentioned here: https://github.com/demisto/etc/issues/33815#issuecomment-785056924

Just want to make sure that it's right before you merge. Thanks

**Comment by Michal-Vardi** [comment]
Done.
@lperryp

**Pull request merged**